File: pie-docs-frontend/src/types/domain/OCR.ts
export type OCRStatus = 'pending' | 'processing' | 'completed' | 'failed' | 'retrying';
export type OCRLanguage = 'ar' | 'en' | 'ar-en' | 'auto';
export type OCRQuality = 'low' | 'medium' | 'high' | 'excellent';

export interface OCRJob {
  id: string;
  documentId: string;
  status: OCRStatus;
  progress: number; // 0-100
  language: OCRLanguage;
  detectedLanguage?: OCRLanguage;
  startTime: string; // ISO string
  endTime?: string; // ISO string
  estimatedTimeRemaining?: number; // seconds
  processingSettings: OCRProcessingSettings;
  retryCount: number;
  maxRetries: number;
  error?: OCRError;
}

export interface OCRProcessingSettings {
  enableLanguageDetection: boolean;
  targetLanguages: OCRLanguage[];
  qualityThreshold: number; // 0-100
  imagePreprocessing: {
    enhanceContrast: boolean;
    denoiseImage: boolean;
    deskewImage: boolean;
    resolutionDPI: number;
  };
  textProcessing: {
    preserveFormatting: boolean;
    extractTables: boolean;
    extractHeaders: boolean;
    mergeFragments: boolean;
  };
}

export interface OCRResult {
  id: string;
  jobId: string;
  documentId: string;
  extractedText: string;
  formattedText?: string;
  language: OCRLanguage;
  confidence: OCRConfidenceScore;
  qualityMetrics: OCRQualityMetrics;
  textBlocks: OCRTextBlock[];
  processingTime: number; // seconds
  dateCreated: string; // ISO string
}

export interface OCRTextBlock {
  id: string;
  text: string;
  confidence: number; // 0-100
  boundingBox: {
    x: number;
    y: number;
    width: number;
    height: number;
  };
  language?: OCRLanguage;
  type: 'paragraph' | 'heading' | 'table' | 'list' | 'other';
  order: number;
}

export interface OCRConfidenceScore {
  overall: number; // 0-100
  character: number; // 0-100
  word: number; // 0-100
  line: number; // 0-100
  paragraph: number; // 0-100
}

export interface OCRQualityMetrics {
  textCoverage: number; // percentage of document with text
  averageWordLength: number;
  punctuationRatio: number;
  specialCharacterRatio: number;
  layoutPreservation: number; // 0-100
  quality: OCRQuality;
  recommendations: string[];
}

export interface OCRError {
  code: string;
  message: string;
  details?: Record<string, unknown>;
  timestamp: string; // ISO string
  recoverable: boolean;
}

export interface OCRQueue {
  jobs: OCRJob[];
  activeJobs: number;
  maxConcurrentJobs: number;
  totalProcessingTime: number; // seconds
  averageProcessingTime: number; // seconds
  isProcessing: boolean;
}

export interface OCRRetryOptions {
  jobId: string;
  newSettings?: Partial<OCRProcessingSettings>;
  reason?: string;
}

export interface OCRPreviewData {
  originalImageUrl?: string;
  preprocessedImageUrl?: string;
  extractedText: string;
  formattedText?: string;
  confidence: OCRConfidenceScore;
  highlightedBlocks: OCRTextBlock[];
  editable: boolean;
}

export interface OCREditSession {
  id: string;
  ocrResultId: string;
  originalText: string;
  editedText: string;
  changes: OCRTextChange[];
  dateStarted: string; // ISO string
  dateLastModified: string; // ISO string
  isCompleted: boolean;
}

export interface OCRTextChange {
  id: string;
  blockId: string;
  originalText: string;
  newText: string;
  changeType: 'correction' | 'addition' | 'deletion' | 'formatting';
  timestamp: string; // ISO string
  confidence?: number;
}

// Component Props Types
export interface OCRProcessorProps {
  documentId: string;
  autoStart?: boolean;
  onComplete?: (result: OCRResult) => void;
  onError?: (error: OCRError) => void;
  onProgress?: (progress: number) => void;
}

export interface OCRStatusIndicatorProps {
  job: OCRJob;
  showDetails?: boolean;
  compact?: boolean;
}

export interface OCRQualityIndicatorProps {
  qualityMetrics: OCRQualityMetrics;
  confidence: OCRConfidenceScore;
  showRecommendations?: boolean;
}

export interface OCRTextPreviewProps {
  ocrResult: OCRResult;
  previewData: OCRPreviewData;
  editMode?: boolean;
  onTextEdit?: (changes: OCRTextChange[]) => void;
  onSave?: (editedText: string) => void;
}

export interface OCRRetryControlsProps {
  job: OCRJob;
  onRetry: (options: OCRRetryOptions) => void;
  onSettingsChange?: (settings: Partial<OCRProcessingSettings>) => void;
  disabled?: boolean;
}

File: pie-docs-backend/app/models/ocr.py
"""
OCR-related Pydantic models
"""
from pydantic import BaseModel
from typing import Optional, List, Dict, Any
from datetime import datetime
from uuid import UUID
from decimal import Decimal


# ==========================================
# OCR Job Models
# ==========================================

class OCRJobBase(BaseModel):
    document_id: UUID
    language: str = "auto"
    max_retries: int = 3


class OCRJobCreate(OCRJobBase):
    pass


class OCRJob(OCRJobBase):
    id: UUID
    status: str  # pending, processing, completed, failed
    progress: int = 0
    detected_language: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    estimated_completion: Optional[datetime] = None
    processing_duration: Optional[int] = None  # seconds
    retry_count: int = 0
    error_code: Optional[str] = None
    error_message: Optional[str] = None
    created_by: Optional[UUID] = None
    created_at: datetime

    class Config:
        from_attributes = True


class OCRJobListResponse(BaseModel):
    jobs: List[OCRJob]
    total: int
    page: int
    page_size: int


# ==========================================
# OCR Result Models
# ==========================================

class OCRResultBase(BaseModel):
    job_id: UUID
    document_id: UUID
    extracted_text: str
    formatted_text: Optional[str] = None
    language: Optional[str] = None
    overall_confidence: Optional[Decimal] = None
    processing_time: Optional[int] = None


class OCRResultCreate(OCRResultBase):
    pass


class OCRResult(OCRResultBase):
    id: UUID
    created_at: datetime

    class Config:
        from_attributes = True


# ==========================================
# OCR Text Block Models
# ==========================================

class OCRTextBlockBase(BaseModel):
    ocr_result_id: UUID
    text: str
    page_number: int
    sequence: int
    confidence: Optional[Decimal] = None
    bounding_box: Optional[Dict[str, Any]] = None  # {x, y, width, height}
    block_type: Optional[str] = None  # paragraph, line, word


class OCRTextBlockCreate(OCRTextBlockBase):
    pass


class OCRTextBlock(OCRTextBlockBase):
    id: UUID

    class Config:
        from_attributes = True


# ==========================================
# OCR Quality Metrics Models
# ==========================================

class OCRQualityMetricsBase(BaseModel):
    ocr_result_id: UUID
    text_coverage: Optional[Decimal] = None
    layout_preservation: Optional[Decimal] = None
    quality_rating: Optional[str] = None  # excellent, good, fair, poor
    issues: List[str] = []
    recommendations: List[str] = []


class OCRQualityMetricsCreate(OCRQualityMetricsBase):
    pass


class OCRQualityMetrics(OCRQualityMetricsBase):
    id: UUID
    created_at: datetime

    class Config:
        from_attributes = True


# ==========================================
# OCR Edit History Models
# ==========================================

class OCREditHistoryBase(BaseModel):
    ocr_result_id: UUID
    original_text: str
    edited_text: str
    change_summary: Optional[str] = None


class OCREditHistoryCreate(OCREditHistoryBase):
    pass


class OCREditHistory(OCREditHistoryBase):
    id: UUID
    user_id: UUID
    created_at: datetime

    class Config:
        from_attributes = True
